File: myFinance_improved_V2/lib/features/sale_product/domain/entities/cart_item.dart
/// Cart item entity - represents a product in the shopping cart
/// Supports variant products from get_inventory_page_v6
class CartItem {
  final String id;
  final String productId;
  final String? variantId;
  final String sku;
  final String name;
  final String? image;
  final double price;
  final int quantity;
  final int available;
  final int customerOrdered;

  const CartItem({
    required this.id,
    required this.productId,
    this.variantId,
    required this.sku,
    required this.name,
    this.image,
    required this.price,
    required this.quantity,
    required this.available,
    this.customerOrdered = 0,
  });

  /// Unique identifier for cart comparison
  /// Uses variantId if exists, otherwise productId
  String get uniqueId => variantId ?? productId;

  double get subtotal => price * quantity;

  CartItem copyWith({
    String? id,
    String? productId,
    String? variantId,
    String? sku,
    String? name,
    String? image,
    double? price,
    int? quantity,
    int? available,
    int? customerOrdered,
  }) {
    return CartItem(
      id: id ?? this.id,
      productId: productId ?? this.productId,
      variantId: variantId ?? this.variantId,
      sku: sku ?? this.sku,
      name: name ?? this.name,
      image: image ?? this.image,
      price: price ?? this.price,
      quantity: quantity ?? this.quantity,
      available: available ?? this.available,
      customerOrdered: customerOrdered ?? this.customerOrdered,
    );
  }
}

File: website/src/features/sale_product/domain/entities/CartItem.ts
/**
 * CartItem Entity
 * Domain entity representing an item in the shopping cart
 * v6: Added variant support
 */

export class CartItem {
  constructor(
    public readonly productId: string,
    public readonly sku: string,
    public readonly productName: string,
    public quantity: number,
    public readonly unitPrice: number,
    public readonly costPrice: number,
    // v6: variant field for unique identification
    public readonly variantId: string | null = null
  ) {}

  // v6: unique key combining product_id + variant_id
  get uniqueKey(): string {
    return this.variantId ? `${this.productId}-${this.variantId}` : this.productId;
  }

  get totalPrice(): number {
    return this.quantity * this.unitPrice;
  }

  get totalCost(): number {
    return this.quantity * this.costPrice;
  }

  incrementQuantity(): CartItem {
    return new CartItem(
      this.productId,
      this.sku,
      this.productName,
      this.quantity + 1,
      this.unitPrice,
      this.costPrice,
      this.variantId
    );
  }

  decrementQuantity(): CartItem | null {
    if (this.quantity <= 1) {
      return null; // Item should be removed from cart
    }
    return new CartItem(
      this.productId,
      this.sku,
      this.productName,
      this.quantity - 1,
      this.unitPrice,
      this.costPrice,
      this.variantId
    );
  }

  updateQuantity(newQuantity: number): CartItem | null {
    if (newQuantity <= 0) {
      return null; // Item should be removed from cart
    }
    return new CartItem(
      this.productId,
      this.sku,
      this.productName,
      newQuantity,
      this.unitPrice,
      this.costPrice,
      this.variantId
    );
  }

  /**
   * Create CartItem from product data
   * v6: Added variantId parameter
   */
  static fromProduct(
    productId: string,
    sku: string,
    productName: string,
    unitPrice: number,
    costPrice: number,
    quantity: number = 1,
    variantId: string | null = null
  ): CartItem {
    return new CartItem(productId, sku, productName, quantity, unitPrice, costPrice, variantId);
  }
}
